### PR TITLE
Initial commit of sysstat version 20151012

### PR DESCRIPTION
**Pull request by jimklimov**
Another useful tool from the author of mbuffer

Note : `sysstat` includes a daemon mode (`sysstatd`) which is currently provided "as is" but without an SMF manifest - this piece is left for another PR if anyone wants to figure out how and why to run it at all, what permissions are needed, etc. ;)

Assumption from glancing over docs is is that multiple hosts may run their daemons and an admin workstation can collect their data over LAN (broadcasts) - this implies privacy, security, firewall and other concerns that are a lot of research out of scope for initial recipe which just adds the utility.

### CODE REVIEW

File: components/sysutils/sysstat/Makefile [hunk 1]
```diff
@@ -0,0 +1,48 @@
+#
```
Comment (pyhalov) [inline components/sysutils/sysstat/Makefile:1]: As always, category better to correspond to fmri

File: components/sysutils/sysstat/Makefile [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+COMPONENT_LICENSE_FILE=	LICENSE
+
+include $(WS_MAKE_RULES)/prep.mk
+include $(WS_MAKE_RULES)/justmake.mk
+include $(WS_MAKE_RULES)/ips.mk
+
+COMPONENT_PRE_CONFIGURE_ACTION =  \
+		($(CLONEY) $(SOURCE_DIR) $(@D))
+
+# non-autoconf "configure" script here
```
Comment (pyhalov) [inline components/sysutils/sysstat/Makefile]: perhaps, use configure.mk and CONFIGURE_OPTIONS =  (I mean, completely overwrite them?)

Comment (jimklimov) [reply components/sysutils/sysstat/Makefile]: Not sure it makes sense here to pretend that the script is compatible to autoconf configure (as other parts of that makefile can assume now or later) when it is an arbitrary script with the same name. As for FMRI - this and xjobs recipes were copied from mbuffer which got into this classification :)

Comment (pyhalov) [reply components/sysutils/sysstat/Makefile]: @xen0l - what do you think about it? I still feel inconvenient about this PRE_BUILD_ACTION. 
Also, as script doesn't use configure.mk, COMPONENT_PRE_CONFIGURE_ACTION is not used.

**Comment by xen0l** [comment]
We should either deliver SMF manifest for sysstatd or drop it for now. Also, separating these into two packages would be good.

File: components/shell/sysstat/Makefile [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#
+
+#
+# Copyright 2016, Jim Klimov
+#
+include ../../../make-rules/shared-macros.mk
+
+COMPONENT_NAME=		sysstat
+COMPONENT_VERSION=	20151012
+COMPONENT_FMRI=		shell/sysstat
```
Comment (xen0l) [inline components/shell/sysstat/Makefile]: shell/sysstat -> diagnostic/sysstat, also ship it in sysutils/

**Comment by xen0l** [comment]
Also, please drop mbuffer changes.

File: components/shell/sysstat/sysstat.p5m [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+set name=info.classification value="$(COMPONENT_CLASSIFICATION)"
+set name=info.upstream-url value="$(COMPONENT_PROJECT_URL)"
+set name=info.source-url value=$(COMPONENT_ARCHIVE_URL)
+set name=org.opensolaris.consolidation value=$(CONSOLIDATION)
+
+license $(COMPONENT_LICENSE_FILE) license='$(COMPONENT_LICENSE)'
+
+file path=usr/bin/$(MACH64)/sysstat
+file path=usr/bin/$(MACH32)/sysstat
+file path=usr/sbin/$(MACH64)/sysstatd
```
Comment (xen0l) [inline components/shell/sysstat/sysstat.p5m]: Separate sysstatd into separate package. I expect people will use sysstat alone.

File: components/shell/sysstat/sysstat.p5m [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+file path=usr/sbin/$(MACH32)/sysstatd
```
Comment (xen0l) [inline components/shell/sysstat/sysstat.p5m]: No manpage for sysstatd?

**Review by xen0l — CHANGES_REQUESTED** [review]

File: components/shell/sysstat/sysstat.p5m (hunk 5 above)
Comment (jimklimov) [reply components/shell/sysstat/sysstat.p5m]: None, and few mentions in the docs. Source file and content (somewhat) is different, though.

File: components/shell/sysstat/sysstat.p5m (hunk 4 above)
Comment (jimklimov) [reply components/shell/sysstat/sysstat.p5m]: It seems (from readme) that `sysstat -d` provides similar functionality. But ok, separated if you like so.

File: components/shell/sysstat/Makefile (hunk 3 above)
Comment (jimklimov) [reply components/shell/sysstat/Makefile]: done

**Comment by jimklimov** [comment]
> We should either deliver SMF manifest for sysstatd or drop it for now. Also, separating these into two packages would be good.

You did not read the PR header, did you? Or chose to ignore it? TL;DR version is: you want a feature - merge the foundation (if there's no fatal flaws) and vote with your PRs on top - this gets around faster, and the commonly satisfactory part of the recipe gets merged earlier and does not rot as part of a dormant PR ;)

**Comment by jimklimov** [comment]
@xen0l : So... how is the latest PR version?

Most of your comments were addressed, except for an SMF service. I am not ready now to look at what it requires to run and how to do it securely, so would rather leave the exercise for somebody who would be first (next) to use this to monitor several OI/Hipster installations from one console ;)

**Comment by xen0l** [comment]
I am fine with the changes, but I don't like shipping a service package without SMF manifest. @pyhalov your take on this?

**Comment by pyhalov** [comment]
If we deliver daemon which can be usable in any way, we should provide SMF manifest.

**Comment by jimklimov** [comment]
At this time I'm not responsibly ready to wrap the service - I am not
certain what it does, what rights it needs and what insecurities it brings
- or carefully does not and so successfully avoids.

We can deliver just the daemon binary, so that somebody else who ever (if
ever) needs this feature, researches and SMFizes it correctly for their
system, and upstreams the change to us. We can also drop delivering the
daemon and stiffle innovation in this area ;)

If you already have a strong opinion about how this should be SMFized -
feel free to merge the PR with code+recipe and follow up with your PR with
a manifest shortly.

2016-11-09 19:11 GMT+01:00 Alexander Pyhalov notifications@github.com:

**Comment by xen0l** [comment]
I don't have time for investigating SMF service right now. Would you be able to provide even basic SMF service? RBAC integration is just an added value in many cases.

**Comment by jimklimov** [comment]
Some manifest added (untested) :)

**Comment by xen0l** [comment]
What does svccfg validate say? Can you at least run the service and application works?

**Comment by pyhalov** [comment]
Did you check service? Can you update Makefile to use configure.mk?

File: components/sysutils/sysstat/Makefile (hunk 2 above)
Comment (jimklimov) [reply components/sysutils/sysstat/Makefile]: Thanks, good catch with this one - removed the action with no losses :)

**Comment by jimklimov** [comment]
Ok, so I experimented with this a bit.

1) The daemon is required in any case - if a service is not running, the client launches it as a sub-process to monitor the local machine, using the same multicast strategy (by default)

2) If the daemon is started manually, or via SMF, the client connects to it
2.1) If I then kill the daemon, client stops refreshing

3) not checked - the client should be able to track a farm reported by sysstatd's on each node.

It seems the impact of networking daemon is moderate: systems can not be managed over the net, only monitored. It does however expose system stats so anyone can subscribe with a sysstat client or even sniff them (unencrypted plaintext), which may be undesirable in shared networks but okay in managed LANs.

The data are sent out as multicasts, about 4kb/sec, it depends on amount of hosts and router/switch support (and imposed limits) whether the traffic is considered large or not, and would reach the monitoring station in same subnet or LAN. Note the packets will likely be fragmented UDP (per comment in code, there was a legacy typo with chunk size not equal to common MTU, and now for backwards-compatibility the project chose to keep it). Probably it won't be an issue on Jumbo VLANs though :)

File: components/sysutils/sysstat/sysstat.p5m [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+# Client part
+file path=usr/bin/$(MACH64)/sysstat
+file path=usr/bin/$(MACH32)/sysstat
+file path=usr/share/man/man1m/sysstat.1m
+
+# Should run the binary matching the kernel bitness
+hardlink path=usr/bin/sysstat target=../lib/isaexec pkg.linted=true
+
+# Server part, required for client even locally, even if not networked via the -l CLI switch
+license $(COMPONENT_LICENSE_FILE) license='$(COMPONENT_LICENSE)'
```
Comment (pyhalov) [inline components/sysutils/sysstat/sysstat.p5m]: Remove second license action

File: components/sysutils/sysstat/Makefile [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+COMPONENT_PROJECT_URL=	http://www.maier-komor.de/sysstat.html
+COMPONENT_LICENSE=	GPLv3
+COMPONENT_LICENSE_FILE=	LICENSE
+
+include $(WS_MAKE_RULES)/prep.mk
+include $(WS_MAKE_RULES)/justmake.mk
+include $(WS_MAKE_RULES)/ips.mk
+
+# non-autoconf "configure" script here
+COMPONENT_PRE_BUILD_ACTION = \
```
Comment (pyhalov) [inline components/sysutils/sysstat/Makefile:37]: Please, convert it to using configure.mk, like nginx.

Comment (jimklimov) [reply components/sysutils/sysstat/Makefile:37]: It is NOT using autoconf configure. So IMHO the recipe integration for configure.mk is a poor choice here. We have many (112 ATM) recipes for projects with justmake.mk, including projects who poorly chose the name for their build-setup scripts "configure" without following the autoconf CLI-API.

Of course, anything can be twisted into anything else (e.g. by overriding proper autoconfigure with pieces for this custom one), but I assume it is a dark path to follow ;)

Conversely: is there something that does not work up to spec with recipe as is? Is there a practical reason to twist it?

File: components/sysutils/sysstat/Makefile (hunk 1 above)
Comment (jimklimov) [reply components/sysutils/sysstat/Makefile:1]: Perhaps, but

1) There is no "diagnostic/" directory
2) There are many packages with "diagnostic/" FMRI, mostly sniffers and local load analyzers
3) The closest equivalent to sysstat is "diagnostic/top" which is why I picked this FMRI patern

Comment (pyhalov) [reply components/sysutils/sysstat/Makefile:1]: I'm fine with this category

File: components/sysutils/sysstat/Makefile (hunk 7 above)
Comment (pyhalov) [reply components/sysutils/sysstat/Makefile:37]: @xenol: Are you OK with it?

Comment (xen0l) [reply components/sysutils/sysstat/Makefile:37]: @pyhalov yes. @jimklimov thanks for providing the testing results.

**Review by xen0l — APPROVED** [review]

**Comment by xen0l** [comment]
This change has been marked for inclusion. Please, squash commits.

**Comment by jimklimov** [comment]
Thanks. Tomorrow, i guess.

--
Typos courtesy of GMail on my Samsung Android

08.12.2016 20:56 пользователь "Adam Števko" <notifications@github.com>
написал:

File: components/sysutils/sysstat/Makefile (hunk 7 above)
Comment (pyhalov) [reply components/sysutils/sysstat/Makefile:37]: Perhaps, at least split it in several (2?) lines to fit in 80 lines?

**Comment by jimklimov** [comment]
Squashed

**Pull request merged**